api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,246 @@
+	// when not set.
+	// +optional
+	CommonName string `json:"commonName,omitempty"`
+
+	// trustDomain is the istio cluster's trust domain, which will also be used for deriving
+	// spiffe URI.
+	// +required
+	TrustDomain string `json:"trustDomain,omitempty"`
+
+	// certificateDNSNames contains the additional DNS names to be added to the istiod certificate SAN.
+	// +listType=set
+	// +optional
+	CertificateDNSNames []string `json:"certificateDNSNames,omitempty"`
+
+	// certificateDuration is the istio-csr and the istiod certificates validity period.
+	// +kubebuilder:default:="1h"
+	// +optional
+	CertificateDuration *metav1.Duration `json:"certificateDuration,omitempty"`
+
+	// certificateRenewBefore is the ahead time to renew the istio-csr and the istiod certificates
+	// before expiry.
+	// +kubebuilder:default:="30m"
+	// +optional
+	CertificateRenewBefore *metav1.Duration `json:"certificateRenewBefore,omitempty"`
+
+	// privateKeySize is the istio-csr and the istiod certificate's key size. When the SignatureAlgorithm
+	// is RSA, must be >= 2048 and for ECDSA, can only be 256 or 384, corresponding to P-256 and P-384 respectively.
+	// +kubebuilder:default:=2048
+	// +kubebuilder:validation:XValidation:rule="oldSelf == 0 || self == oldSelf",message="privateKeySize is immutable once set"
+	// +optional
+	PrivateKeySize int `json:"privateKeySize,omitempty"`
+
+	// signatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="signatureAlgorithm is immutable once set"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration *metav1.Duration `json:"maxCertificateDuration,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for obtaining the certificates.
+type ServerConfig struct {
+	// port to serve istio-csr gRPC service.
+	// +kubebuilder:default:=443
+	// +kubebuilder:validation:XValidation:rule="oldSelf == 0 || self == oldSelf",message="port is immutable once set"
+	// +optional
+	Port int32 `json:"port,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +listType=atomic
+	// +kubebuilder:default:={"default"}
+	// +kubebuilder:validation:XValidation:rule="self.all(x, x in oldSelf) && oldSelf.all(x, x in self)",message="revisions is immutable once set"
```

Comment:
not sure if I read this correctly, does it want to make sure that old revisions can never get removed once added?
What happens when the list goes beyond MaxItems=10, how do we purge it from apiserver capabilities with this CEL enabled.

Reply:
Yes, revisions can only be configured when creating `istiocsr` resource and cannot be modified later.
With `MaxItems` set to 10, a user cannot configure more than that, creating the resource would fail when it crosses the defined limit.

Reply:
> Are there UTs added for testing the CEL validations? If not, it's better to add them to verify the validations.

Yup, we should add something for the validations (seems this is the first time we're adding CEL based validation to our operator API) based upon `envtest`, to begin with: https://book.kubebuilder.io/reference/envtest#writing-tests sorta. 

For reference, the more exhaustive related tests we've in openshift/api: https://github.com/openshift/api/blob/188620a9ab5eaad36199cf83d9a0afcb08fdfd52/tests/crd_filter.go + https://github.com/openshift/api/blob/188620a9ab5eaad36199cf83d9a0afcb08fdfd52/machineconfiguration/v1alpha1/tests/machineconfignodes.machineconfiguration.openshift.io/MachineConfigNodes.yaml#L113-L158, ofcourse we want something functional and much more minimal than that.

Reply:
> Yes, revisions can only be configured when creating `istiocsr` resource and cannot be modified later. With `MaxItems` set to 10, a user cannot configure more than that, creating the resource would fail when it crosses the defined limit.

Alright, so my understanding is that user sets these revisions (and not operator added)
- never let user remove an existing one, only append is allowed;
- we're blocking >10.